cpp/src/lookup/lookup_src_dst_impl.cuh

```diff
@@ -0,0 +1,805 @@
+
+    thrust::reduce_by_key(handle.get_thrust_policy(),
+                          std::get<0>(unique_pairs_buffer).begin(),
+                          std::get<0>(unique_pairs_buffer).end(),
+                          unique_pair_counts.begin(),
+                          unique_types.begin(),
+                          unique_type_counts.begin());
+
+  } else {
+    auto edge_types = cugraph::extract_transform_e(
+      handle,
+      graph_view,
+      cugraph::edge_src_dummy_property_t{}.view(),
+      cugraph::edge_dst_dummy_property_t{}.view(),
+      edge_type_view,
+      cuda::proclaim_return_type<thrust::optional<edge_type_t>>(
+        [] __device__(auto, auto, thrust::nullopt_t, thrust::nullopt_t, edge_type_t et) {
+          return thrust::optional<edge_type_t>{et};
+        }));
+
+    thrust::sort(handle.get_thrust_policy(), edge_types.begin(), edge_types.end());
+
+    auto nr_unique_types =
+      thrust::count_if(handle.get_thrust_policy(),
+                       thrust::make_counting_iterator(size_t{0}),
+                       thrust::make_counting_iterator(edge_types.size()),
+                       detail::is_first_in_run_t<edge_type_t const*>{edge_types.data()});
+
+    unique_types.resize(static_cast<size_t>(nr_unique_types), handle.get_stream());
+    unique_type_counts.resize(static_cast<size_t>(nr_unique_types), handle.get_stream());
+
+    thrust::reduce_by_key(handle.get_thrust_policy(),
+                          edge_types.begin(),
+                          edge_types.end(),
+                          thrust::make_constant_iterator(size_t{1}),
+                          unique_types.begin(),
+                          unique_type_counts.begin());
+  }
+
+  std::vector<edge_type_t> h_unique_types(unique_types.size());
+  std::vector<edge_t> h_unique_type_counts(unique_types.size());
+
+  raft::update_host(
+    h_unique_types.data(), unique_types.data(), unique_types.size(), handle.get_stream());
+
+  raft::update_host(h_unique_type_counts.data(),
+                    unique_type_counts.data(),
+                    unique_type_counts.size(),
+                    handle.get_stream());
+
+  handle.sync_stream();
+
+  auto search_container =
+    EdgeTypeAndIdToSrcDstLookupContainerType(handle, h_unique_types, h_unique_type_counts);
+
+  //
+  // Populate the search container
+  //
+
+  for (size_t local_ep_idx = 0; local_ep_idx < graph_view.number_of_local_edge_partitions();
+       ++local_ep_idx) {
+    //
+    // decompress one edge_partition at a time
+    //
+
+    auto number_of_local_edges =
+      graph_view.local_edge_partition_view(local_ep_idx).number_of_edges();
+
+    if (graph_view.has_edge_mask()) {
+      number_of_local_edges =
+        detail::count_set_bits(handle,
+                               (*(graph_view.edge_mask_view())).value_firsts()[local_ep_idx],
+                               number_of_local_edges);
+    }
```

Comment:
Use this instead.

https://github.com/rapidsai/cugraph/blob/branch-24.08/cpp/include/cugraph/edge_partition_device_view.cuh#L611

Reply:
Changed.